### PR TITLE
User management java sdk implementation

### PR DESCRIPTION
**Pull request by suganyakrishnan**
Java SDK implementation for
1. Users routes
2. Policies routes
3. Groups routes

### CODE REVIEW

File: src/main/java/com/spotinst/sdkjava/SpotinstClient.java [hunk 1]
```diff
@@ -117,6 +117,11 @@ public static SpotinstAccountAdminClient getAdminAccountClient(String authToken)
         return spotAdminAccountClient;
     }
 
+    public static SpotinstAccountAdminClient getAdminAccountClient(String authToken, String account) {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/SpotinstClient.java:120]: Will this be another client, together with existing client?

File: src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+    public static void main(String[] args) {
+        SpotinstAccountAdminClient adminClient = SpotinstClient.getAdminAccountClient(authToken, accountId);
+
+        // User APIs
+        String userId = createUser(adminClient);
+        getUserDetails(adminClient, userId);
+        String programmaticUser = createProgammaticUser(adminClient);
+        getUserDetails(adminClient, programmaticUser);
+        getOrganizationUsers(adminClient);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java:25]: Need to understand why organization user API, needs account ID.

File: src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        updateGroupsofUsers(adminClient, userId);
+        updateUserDirectPolicies(adminClient, userId);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java:27]: deleteUser API is not there?

File: src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        UserDetails userDetails = adminClient.getUserDetails(userId);
+        System.out.println(String.format("userName: %s", userDetails.getUsername()));
+        System.out.println(String.format("type: %s", userDetails.getType()));
+        System.out.println(String.format("mfa: %s", userDetails.getMfa()));
+    }
+
+    private static String createProgammaticUser(SpotinstAccountAdminClient adminClient) {
+        UserPolicies.Builder policiesBuilder  = UserPolicies.Builder.get();
+        UserPolicies         userPolicies     = policiesBuilder.setPolicyId("4").build();
+        List<UserPolicies>   userPoliciesList = Collections.singletonList(userPolicies);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java:66]: can we do userPoliciesList.add(userPolicies)?

File: src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        // Build user
+        UserDetails userDetails = adminClient.getUserDetails(userId);
+        System.out.println(String.format("userName: %s", userDetails.getUsername()));
+        System.out.println(String.format("type: %s", userDetails.getType()));
+        System.out.println(String.format("mfa: %s", userDetails.getMfa()));
+    }
+
+    private static String createProgammaticUser(SpotinstAccountAdminClient adminClient) {
+        UserPolicies.Builder policiesBuilder  = UserPolicies.Builder.get();
+        UserPolicies         userPolicies     = policiesBuilder.setPolicyId("4").build();
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java:65]: Change class name to UserPolicy

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import com.spotinst.sdkjava.client.rest.IPartialUpdateEntity;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiOrganizationUserGroups implements IPartialUpdateEntity {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java:16]: Didn't get this class.

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiPolicyContent.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.List;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiPolicyContent implements IPartialUpdateEntity {
+    @JsonIgnore
+    private Set<String>            isSet;
+    private List<ApiPolicyStatements> statements;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiPolicyContent.java:19]: Please change it to ApiPolicyStatement

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiProgrammaticUser.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiProgrammaticUser implements IPartialUpdateEntity {
+    @JsonIgnore
+    private Set<String>           isSet;
+    private String                name;
+    private String                description;
+    private List<ApiUserPolicies> policies;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiProgrammaticUser.java:21]: Please change it to ApiUserPolicy

File: src/main/java/com/spotinst/sdkjava/SpotinstClient.java (hunk 1 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/SpotinstClient.java:120]: Yes. It is another client which can be used for account level APIs

File: src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java (hunk 3 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java:27]: Yes. DeleteUser is not supported through API with New/Old UMS

File: src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java (hunk 5 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/example/admin/account/UserManagementUsageExample.java:65]: Updated

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java (hunk 6 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java:16]: This class is to get all groups in the organization.

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiPolicyContent.java (hunk 7 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiPolicyContent.java:19]: Renamed

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiProgrammaticUser.java (hunk 8 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiProgrammaticUser.java:21]: Renamed

File: src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    private              String name;
+
+    private UserTypeEnum(String name) {
+        this.name = name;
+    }
+
+    public static UserTypeEnum fromName(String name) {
+        UserTypeEnum retVal = null;
+
+        for (UserTypeEnum policyTypeEnum : UserTypeEnum.values()) {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java:21]: Variable name should be changed.

File: src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package com.spotinst.sdkjava.enums.admin.account;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public enum UserTypeEnum {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java:6]: Which API is using this enum?

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    public String getCreatedAt() {
+        return createdAt;
+    }
+
+    public void setCreatedAt(String createdAt) {
+        isSet.add("createdAt");
+        this.createdAt = createdAt;
+    }
+
+    public String getUsersCount() {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java:74]: Is this a string?

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUpdateDirectPoliciesRequest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import com.spotinst.sdkjava.client.rest.IPartialUpdateEntity;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiUpdateDirectPoliciesRequest implements IPartialUpdateEntity {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUpdateDirectPoliciesRequest.java:16]: Rename class to ApiUpdatePoliciesRequest

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserPolicy.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import com.spotinst.sdkjava.client.rest.IPartialUpdateEntity;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiUserPolicy implements IPartialUpdateEntity {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserPolicy.java:16]: ApiUserDetailsPolicy

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserTokens.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.spotinst.sdkjava.client.rest.IPartialUpdateEntity;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiUserTokens implements IPartialUpdateEntity {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserTokens.java:15]: ApiUserDetailsTokens

File: src/main/java/com/spotinst/sdkjava/model/bl/admin/account/UpdateDirectPoliciesRequest.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class UpdateDirectPoliciesRequest {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/admin/account/UpdateDirectPoliciesRequest.java:13]: rename to UpdatePoliciesRequest

File: src/main/java/com/spotinst/sdkjava/model/converters/admin/account/AccountAdminConverter.java [hunk 16]
```diff
@@ -0,0 +1,600 @@
+package com.spotinst.sdkjava.model.converters.admin.account;
+
+import com.spotinst.sdkjava.model.api.admin.account.*;
+import com.spotinst.sdkjava.model.bl.admin.account.*;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AccountAdminConverter {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/converters/admin/account/AccountAdminConverter.java:9]: Class name should be OrganizationAdminConverter

File: src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java (hunk 9 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java:21]: Updated

File: src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java (hunk 10 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/enums/admin/account/UserTypeEnum.java:6]: UserDetails.java and OrganizationUsers.java are using this enum

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java (hunk 11 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiOrganizationUserGroups.java:74]: It is an integer. Updated both bl and api class.

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUpdateDirectPoliciesRequest.java (hunk 12 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUpdateDirectPoliciesRequest.java:16]: Updated both BL and API class

[the remark this answers is not in the document]
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUpdateDirectPoliciesRequest.java:16]: Added

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserPolicy.java (hunk 13 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserPolicy.java:16]: Updated

File: src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserTokens.java (hunk 14 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/api/admin/account/ApiUserTokens.java:15]: Added

File: src/main/java/com/spotinst/sdkjava/model/converters/admin/account/AccountAdminConverter.java (hunk 16 above)
Comment (suganyakrishnan) [reply src/main/java/com/spotinst/sdkjava/model/converters/admin/account/AccountAdminConverter.java:9]: Updated

**Comment by anuragsharma-123** [comment]
The changes in this PR are incorporated via another PR.

**Pull request closed without merging**